CONTRIBUTING.md

```diff
@@ -78,96 +85,160 @@ installed.
+To learn how to set yourself up with a BinderHub development environment that lets
+you modify the builder and launcher refer to the [Python server changes](#python-server-changes) section.
 
-## Develop Kubernetes integration
 
-This requires `helm` and a functional Kubernetes cluster. Please do
-[preliminary Kubernetes setup](#kubernetes-setup) if you haven't already
-before continuing here.
+## Python server changes
 
-With a Kubernetes cluster running, as you verify with `kubectl version`, you can
-continue.
+Setting yourself up to make changes to the kubernetes integration of BinderHub
+requires a few one-time setup steps. These steps are described in the
+"One-time installation" section below. Follow those first then return here for
+day to day development procedures.
+
+
+### Day to day development tasks
+
+After completing the "One-time installation" steps you are ready for day-to-day
+development. These are the tasks you need to perform for every day development:
+
+* Start and stop minikube with `minikube start` and `minikube stop`.
+* Install JupyterHub in minikube with helm `./testing/local-binder-k8s-hub/install-jupyterhub-chart`
+* Setup `docker` to use the same Docker daemon as your minikube cluster `eval $(minikube docker-env)`
+* Start BinderHub `python3 -m binderhub -f testing/local-binder-k8s-hub/binderhub_config.py`
+* Visit your BinderHub at http://localhost:8585
 
-1. Locally install BinderHub as a Python package and its development
-   requirements locally.
+To test that everything works as it should visit http://localhost:8585/v2/gh/binder-examples/minimal-dockerfile/HEAD. This will build a very simple repository and should eventually
+show you the classic Jupyter notebook interface.
+
+To execute most of our test suite run: `pytest -svx -m "not auth"`
+
+The tests should generate familiar pytest like output and complete in a few
+minutes.
+
+To run more advanced test or the full test suite we recommend you open a Pull Request
+and let our Continuous Integration system run them. This is by far the easiest
+way to find out if your changes pass all tests. Running these tests locally
+is considered an expert task. To find out how to do it please read
+`.github/workflows/test.yml` and replicate it for your local setup.
+
+After working on BinderHub you can [clean up](#cleaning-up-resources).
+
+
+### One-time installation
+
+To run a full BinderHub it needs to have access to a kubernetes cluster
+with a JupyterHub installed on it. This is what we will setup in this section.
+All the steps are given as command-line commands for Ubuntu systems. They are
+used as a common denominator that can be translated into the correct commands
+on your local system.
+
+Before you begin, there are a few utilities that need to be installed:
+```bash
+sudo apt install python3 python3-pip npm curl
+```
+
+If you are on linux, you may additionally need to install socat for port forwarding:
+
+```bash
+sudo apt install socat
+```
```

Comment:
Was socat still needed?

Reply:
I don't have a linux system so I have no idea :-/

Reply:
I removed it after having testing things on my linux system and finding everything to work fine without it installed if I recall correctly. It was at least intentionally removed by me after investigation rather than a mistake.

Reply:
Yeah, you don't need socat anymore